corehq/apps/userreports/models.py

```diff
@@ -206,14 +227,40 @@ def save(self, **params):
     def data_source_id(self):
         return self._id
 
-    def filter(self, document):
+    def filter(self, document, eval_context=None):
+        if eval_context is None:
+            EvaluationContext(document)
+
         filter_fn = self._get_main_filter()
-        return filter_fn(document, EvaluationContext(document, 0))
+        return filter_fn(document, eval_context)
 
     def deleted_filter(self, document):
         filter_fn = self._get_deleted_filter()
         return filter_fn and filter_fn(document, EvaluationContext(document, 0))
 
+    @property
+    def has_validations(self):
+        return len(self.validations) > 0
+
+    def validate_document(self, document, eval_context=None):
+        if eval_context is None:
+            eval_context = EvaluationContext(document)
+
+        for validation in self._validations():
+            if validation.validation_function(document, eval_context) is False:
+                raise ValidationError(validation.name, validation.error_message)
```

Comment:
What do you think about performing all validations and raising an error with all failures rather than just the first one?

Reply:
https://github.com/dimagi/commcare-hq/pull/23799/commits/2718349fda331c29bdbd81d3c647462591326fe7 & https://github.com/dimagi/commcare-hq/pull/23799/commits/bed8edee487028c0f4361d38dd9726b7cc5fe23e